### PR TITLE
Return `distiset` from `Pipeline.run`

### LINKED ISSUE #373: `Pipeline.run` method returning `Distiset` instance
Update the `Pipeline.run` method to return a dataset that can be easily pushed to the Hugging Face Hub. As `distilabel` pipelines can have several leaf steps, that means that the output of each leaf step will generate a dataset i.e. a different config of the dataset.

### PR DESCRIPTION
**Pull request by plaguss**
## Description

This PR creates a `Distiset` class which is a wrapper around a dictionary containing the internal `datasets.Dataset` generated during the `Pipeline.run` method. Each key corresponds to the `leaf_step` name in internal `DAG`, and each value is a `datasets.Dataset`. It has two methods:

- `push_to_hub`: to push the `Distiset` to the hub, where each configuration corresponds to one of the subsets`
- `train_test_split`: which transforms each one of the internal `datasets.Dataset` to a `datasets.DatasetDict` (all the subsets with the same train/test sizes.

The `Pipeline.run` method after finishing will locate in the cache folder the (parquet) files written via `_WriteBuffer` and generate the `Distiset`.

Dummy example:

```python
with Pipeline() as pipeline:
    load_hub_dataset = LoadHubDataset(name="load_dataset", batch_size=8)
    rename_columns = RenameColumns(name="rename_columns", input_batch_size=12)
    generate_response = GenerateResponse(
        name="generate_response", input_batch_size=16
    )

    load_hub_dataset.connect(rename_columns)
    rename_columns.connect(generate_response)

    ds = pipeline.run(
        parameters={
            "load_dataset": {
                "repo_id": "plaguss/test",
                "split": "train",
            },
            "rename_columns": {
                "rename_mappings": {
                    "prompt": "instruction",
                },
            },
        }
    )
# >>> ds["generate_response"]
# Distiset({"leaf_step_1": Dataset(...), "leaf_step_2": Dataset(...)})
```

Closes #373

### CODE REVIEW

File: src/distilabel/pipeline/local.py [hunk 1]
```diff
@@ -285,41 +294,122 @@ def signal_handler(signumber: int, frame: Any) -> None:
...
+        self._buffers = buffs
+
+    def close(self) -> None:
+        for writer in self._writers.values():
+            writer.close()
+
+    # def _combine_batches(self) -> Iterator[Dict[str, Any]]:
+    #     for _, data in self._buffers.items():
+    #         print("DATA", data)
+    #         yield data[-1]
```
Comment (alvarobartt) [inline src/distilabel/pipeline/local.py:412]: I think this is a debugging leftover

Comment (plaguss) [reply src/distilabel/pipeline/local.py:412]: totally

File: src/distilabel/utils/data.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# limitations under the License.
+
+from pathlib import Path
+
+from datasets import DatasetDict, load_dataset
+from pyarrow.lib import ArrowInvalid
+
+from distilabel.utils.logging import get_logger
+
+logger = get_logger("data")
```
Comment (alvarobartt) [inline src/distilabel/utils/data.py:22]: Maybe?

```suggestion
logger = get_logger("utils.data")
```

File: src/distilabel/pipeline/local.py [hunk 3]
```diff
@@ -285,41 +294,122 @@ def signal_handler(signumber: int, frame: Any) -> None:
...
     def __init__(self, path: "PathLike", leaf_steps: Set[str]) -> None:
-        path = Path(path)
-        if path.suffix == "":
-            path = path / "data.jsonl"
-        self._path = path
+        """
+        Args:
+            path (PathLike): Folder where the files will be written, the idea
+                is for this path to be in the cache folder under /data.
+            leaf_steps (Set[str]): Leaf steps from the dag of the pipeline.
```
Comment (alvarobartt) [inline src/distilabel/pipeline/local.py:315]: We agreed on removing the types out of the docstrings as those will be automatically retrieved from the function signature 👍🏻

File: src/distilabel/pipeline/local.py [hunk 4]
```diff
@@ -285,41 +294,122 @@ def signal_handler(signumber: int, frame: Any) -> None:
...
+        self._writers: Dict[str, pq.ParquetWriter] = {}
 
-    @property
-    def is_full(self) -> bool:
-        return all(self._buffers.values())
+    def _get_filename(self, step_name: str) -> str:
+        """Creates the filename for the step.
+
+        Args:
+            step_name (str): Name of the step to which the data pertains.
```
Comment (alvarobartt) [inline src/distilabel/pipeline/local.py:332]: Same here

```suggestion
            step_name: Name of the step to which the data belongs to.
```

File: src/distilabel/pipeline/local.py (hunk 3 above)
Comment (alvarobartt) [reply src/distilabel/pipeline/local.py:315]:
```suggestion
            path: Folder where the files will be written, the idea
                is for this path to be in the cache folder under /data.
            leaf_steps: Leaf steps from either the DAG of the Pipeline.
```

File: tests/integration/test_pipe_simple.py [hunk 5]
```diff
@@ -60,37 +61,40 @@ def outputs(self) -> List[str]:
...
+            name="generate_response", input_batch_size=16
+        )
+
+        load_hub_dataset.connect(rename_columns)
+        rename_columns.connect(generate_response)
+
+        return pipeline.run(
+            parameters={
+                "load_dataset": {
+                    "repo_id": "plaguss/test",
```
Comment (alvarobartt) [inline tests/integration/test_pipe_simple.py:78]: Unrelated to this PR, but maybe it's time to create `argilla-internal-testing` in Hugging Face Hub? 😆

**Review by alvarobartt — COMMENTED** [review]
LGTM so far, but still need to test it further!

File: tests/integration/test_pipe_simple.py (hunk 5 above)
Comment (plaguss) [reply tests/integration/test_pipe_simple.py:78]: yes... the current workflow is weird at best 😆

File: src/distilabel/pipeline/local.py [hunk 6]
```diff
@@ -285,41 +294,130 @@ def signal_handler(signumber: int, frame: Any) -> None:
 
 
 class _WriteBuffer:
```
Comment (gabrielmbmb) [inline src/distilabel/pipeline/local.py:296]: We can move this class to `base.py`

File: src/distilabel/utils/data.py
Comment (gabrielmbmb) [inline src/distilabel/utils/data.py:1]: I would rename this to `src/distilabel/dataset.py` or `distiset.py`

**Review by gabrielmbmb — APPROVED** [review]
LGTM! We can rename some files, but looks good :)

**Pull request merged**